### PR TITLE
Parameterize server settings & remove TestData

### PR DESCRIPTION
**Pull request by JonathanLorimer**
- Minimal implementation of server parameterization [issue here](https://github.com/JonathanLorimer/weft/issues/7)
- Moved all TestData imports in ParserSpec into the ParserSpec file. Not that pretty, but allowed us to get rid of TestData
- Created an example file and executable for all our server tests and checking out the developer UX, this will hopefully turn into an actual example for documentation at some point.

### CODE REVIEW

File: src/Weft/Server.hs [hunk 1]
```diff
@@ -55,8 +52,15 @@ successResponse = responseLBS status200 [(hContentType, "application/json")] . e
 errorResponse :: BL.ByteString -> Response
 errorResponse = responseLBS status500 [(hContentType, "application/json")]
 
-main :: IO ()
-main = do
-    let port = 3000
-    Prelude.putStrLn $ "Listening on port " ++ show port
-    run port $ app gqlResolver -- TODO: allow the user to pass in their resolvers, or Reader them
+server :: (ToJSON (q 'Response), Wefty q)
```
Comment (isovector) [inline src/Weft/Server.hs]: Might as well stick `ToJSON (q 'Response)` into `Wefty q`

File: src/Weft/Server.hs [hunk 2]
```diff
@@ -55,8 +52,15 @@ successResponse = responseLBS status200 [(hContentType, "application/json")] . e
...
-    let port = 3000
-    Prelude.putStrLn $ "Listening on port " ++ show port
-    run port $ app gqlResolver -- TODO: allow the user to pass in their resolvers, or Reader them
+server :: (ToJSON (q 'Response), Wefty q) 
+       => [Settings -> Settings]
+       -> Gql q () () 'Resolver
+       -> IO ()
+server s r = runSettings (L.foldl' (&) defaultSettings s) (app r)
+
+-- main :: IO ()
```
Comment (isovector) [inline src/Weft/Server.hs]: nit: please delete; source control will keep it around if we need it!

File: test/ParserSpec.hs [hunk 3]
```diff
@@ -1,19 +1,54 @@
...
 import           Text.PrettyPrint.HughesPJ
 import           Weft.Generics.PprQuery
 import           Weft.Generics.QueryParser
 import           Weft.Internal.Types
 import           Weft.Types
 
+newtype Id = Id String deriving (Generic, Show, Read, Eq, Ord, Arbitrary, ToJSON)
+newtype Name = Name String deriving (Generic, Show, Eq, Ord, Arbitrary, ToJSON)
+
+data User ts = User
```
Comment (isovector) [inline test/ParserSpec.hs:22]: unrelated to your PR, just thinking aloud:

Gee this is a lot of boilerplate just to model a simple schema. Big constant factors for all the `Magic` machinery, and then all of the standalone deriving crap. What do you think if we encoded `User` as this instead:

```haskell
data User = User
  { userId :: WithArgs UserIdArgs Id
  , userName :: Name
  , userBestFriend :: WithArgs UserBestFriendArgs User
  , userFriends :: [User]
  } deriving (Generic, Show, Eq, ToJSON, Arbitrary)
```

and then replace `User 'Data` with `HKD User Bare` (we could come up with a friendlier looking type, eg `RecordOf User`).

File: src/Weft/Server.hs [hunk 4]
```diff
@@ -55,8 +52,15 @@ successResponse = responseLBS status200 [(hContentType, "application/json")] . e
...
-main :: IO ()
-main = do
-    let port = 3000
-    Prelude.putStrLn $ "Listening on port " ++ show port
-    run port $ app gqlResolver -- TODO: allow the user to pass in their resolvers, or Reader them
+server :: (ToJSON (q 'Response), Wefty q) 
+       => [Settings -> Settings]
+       -> Gql q () () 'Resolver
+       -> IO ()
+server s r = runSettings (L.foldl' (&) defaultSettings s) (app r)
```
Comment (isovector) [inline src/Weft/Server.hs]: This `[Settings -> Settings]` stuff feels like you're looking for the `Endo` monoid.

`L.foldl' (&) defaultSettings s` could instead be `appEndo (foldMap Endo s) defaultSettings`, or even `ala Endo foldMap s defaultSettings` if you're feeling cheeky.

**Review by isovector — APPROVED** [review]

File: test/ParserSpec.hs (hunk 3 above)
Comment (JonathanLorimer) [reply test/ParserSpec.hs:22]: That sounds way better

File: src/Weft/Server.hs (hunk 4 above)
Comment (JonathanLorimer) [reply src/Weft/Server.hs]: @isovector programming with you brings out the cheekiness in me

**Pull request merged**